qiskit_ibm_provider/transpiler/passes/scheduling/block_base_padder.py

```diff
@@ -184,11 +219,216 @@ def _pad(
+            if last_node is not None:
+                last_node_in_block = True
+                try:
+                    self._block_dag.node(last_node._node_id)
+                except IndexError:
+                    last_node_in_block = False
+            else:
+                last_node_in_block = False
+
+            if not (
+                last_node_in_block
+                and isinstance(last_node.op, Measure)
+                and set(node.qargs) == set(last_node.qargs)
+            ):
+                return False
+
+        # Fast path contents are limited to gates and delays
+        for block in node.op.blocks:
+            block_dag = circuit_to_dag(block)
+            for node_ in block_dag.topological_op_nodes():
+                if not isinstance(node_.op, (Gate, Delay)):
+                    return False
+        return True
+
+    def _visit_control_flow_op(self, node: DAGNode) -> None:
+        """Visit a control-flow node to pad."""
+
+        # Control-flow terminator ends scheduling of block currently
+        block_idx, t0 = self._node_start_time[node]  # pylint: disable=invalid-name
+        self._terminate_block(t0, block_idx)
+        self._add_block_terminating_barrier(block_idx, t0, node)
+
+        # Only pad non-fast path nodes
+        fast_path_node = node in self._fast_path_nodes
+
+        # TODO: This is a hack required to tie nodes of control-flow
+        # blocks across the scheduler and block_base_padder. This is
+        # because the current control flow nodes store the block as a
+        # circuit which is not hashable. For processing we are currently
+        # required to convert each circuit block to a dag which is inefficient
+        # and causes node relationships stored in analysis to be lost between
+        # passes as we are constantly recreating the block dags.
+        # We resolve this here by extracting the cached dag blocks that were
+        # stored by the scheduling pass.
+        new_node_block_dags = []
+        for block_idx, _ in enumerate(node.op.blocks):
+            block_dag = self._node_block_dags[node][block_idx]
+            new_node_block_dags.append(
+                self._visit_block(block_dag, pad_wires=not fast_path_node)
+            )
+
+        # Build new control-flow operation containing scheduled blocks
+        # and apply to the DAG.
+        new_control_flow_op = node.op.replace_blocks(
+            dag_to_circuit(block) for block in new_node_block_dags
+        )
+        # Enforce that this control-flow operation contains all wires since it has now been padded
+        # such that each qubit is scheduled within each block. Don't added all cargs as these will not
+        # be padded.
+        self._apply_scheduled_op(
+            block_idx,
+            t0,
+            new_control_flow_op,
+            node.qargs if fast_path_node else self._block_dag.qubits,
+            node.cargs,
+        )
```

Comment:
If they're all now full-width in terms of qubits, is the terminating barriers thing required?  By applying them to all qubits, you've already effectively guaranteed that the block is locked in its current place in the topological order.

Reply:
Yes, because if a block is not fully padded out (such as the fast-path case) the topological ordering will behave incorrectly. I tried to limit barriers in this revision to only where they are necessary.